tests/rptest/tests/tombstone_removal_e2e_test.py

```diff
@@ -0,0 +1,283 @@
+
+        # produce a lot of data records for keys 0-9
+        self.produce_data(10)
+
+        # produce tombstones for keys 5-9 (one for each)
+        tombstone_offsets = [
+            self.rpk.produce(
+                topic=self.topic_name,
+                key="key-{}".format(key),
+                msg="",
+                tombstone=True,
+            )
+            for key in range(5, 10)
+        ]
+        offset_after_tombstones = max(tombstone_offsets) + 1
+
+        # produce a lot of data records for keys 0-4
+        self.produce_data(5)
+
+        # make node 1 the leader
+        self.transfer_leadership_to_node(1)
+
+        # init a replica on node 4
+        self.admin.set_partition_replicas(
+            self.topic_name,
+            0,
+            [
+                {"node_id": n, "core": PartitionMovementMixin.INVALID_CORE}
+                for n in [1, 2, 4]
+            ],
+        )
+        wait_until(
+            lambda: len(self.admin.list_reconfigurations(node=node1)) > 0,
+            timeout_sec=10,
+            backoff_sec=1,
+            err_msg="timed out waiting for reconfiguration to start",
+        )
+
+        # let node 4 recover at least some data but no tombstones
+        def data_recovered_halfway() -> bool:
+            reconfigurations = self.admin.list_reconfigurations(node=node1)
+            self.redpanda.logger.debug(f"reconfigurations: {reconfigurations}")
+            if not reconfigurations:
+                return False
+            bytes_moved = reconfigurations[0]["bytes_moved"]
+            total_bytes_to_move = self.data_msg_count * self.data_msg_size
+            assert bytes_moved <= total_bytes_to_move * 0.75, "node recovered too fast"
+            return bytes_moved >= total_bytes_to_move * 0.25
+
+        wait_until(
+            data_recovered_halfway,
+            timeout_sec=120,
+            backoff_sec=1,
+            err_msg="timed out waiting for partial recovery",
+        )
+
+        with self.redpanda.paused_node(node4):
+            # turn on compaction
+            self.rpk.alter_topic_config(self.topic_name, "cleanup.policy", "compact")
+            self.rpk.alter_topic_config(
+                self.topic_name, "min.cleanable.dirty.ratio", "0.0"
+            )
+
+            # make sure node 1 is not compacting away tombstones until node 4
+            # catches up and compacts away correspondent data records first
+            time.sleep(30)
+            abnormal_compaction = self.no_more_than_n_records_below_offset(
+                0, offset_after_tombstones
+            )
+            if abnormal_compaction:
+                self.logger.error(
+                    "coordinated compaction did not prevent tombstone deletion on node 1"
+                )
+
+        # compact node 1
+        wait_until(
+            lambda: self.no_more_than_n_records_below_offset(
+                10, offset_after_tombstones
```

Comment:
Should this assert for "no more than 5"? You wrote additional keys for 0-4 after tombstones so they will get compacted I guess?